### DISCUSSION TITLE
Bulk Update Content Items/Database

### DISCUSSION THREAD

**Discussion by akim-green** [OrchardCMS/OrchardCore#8279] [Q&A] [answered]
Hello all. I am looking for some advice regarding updating thousands of content items in an interval. The initial code I had worked fine: using a workflow timer to trigger an update via http endpoint every ten minutes.

It stopped working as I added significantly more content items (items in a store, with their price and quantity), the update seems to be taking too long.

So I tried updating batches in parallel using async and When.All, but I started getting database errors regarding accessing the reader while it's busy.

I was wondering if this would be fixed by moving to SQL Server with MARS support, or if there was some way I could do a batch update to the database rather than using the ContentManager.

Any advice would be appreciated.

```
//old
foreach (ContentItem store in stores)
     retval += await contentService.UpdateStoreInventory(store)

//new
foreach (ContentItem store in stores)
{
     taskList.Add(contentService.UpdateStoreInventory(store));
 }
await Task.WhenAll(taskList);
```

**Comment by deanmarcussen** [comment]
> await Task.WhenAll(taskList); 

Won't work. A YesSql session is not threadsafe, so can't be used like this. MARS won't help.

Some recent updates to `YesSql` include better batching, and a reduction in sql queries during the update.

That being said updating thousands on content items is still going to be a lot of work for the content manager / YesSql.

Depending on what values you are updating @Skrypt might suggest some custom JSON SQL. But it really depends on

- what values you are updating on these thoasands of content items, and why.
- Whether a content item is really the most suitable model for an inventory record.

**Reply by akim-green** [reply]
Thank you @deanmarcussen this clarified a lot for me! I opted for using a content item mainly because I was comfortable with the content manager and due to front requirements to get product by price or inventory via SQL. However the inventory is totally dictated by data from an external api and the values need not be modified in the CMS, so I guess I don't strictly need a content item, I'll look into an alternate model.

**Comment by Skrypt** [comment]
It would be relatively simple to create a tool that translate a SQL Server JSON Function to other DB SQL syntax.

**Reply by akim-green** [reply]
Thanks @Skrypt, I am not too familiar with the JSON functionality in SQL Server, but I will definitely check it out and see what I can accomplish, really appreciate the advice!

**Reply by Skrypt** [reply]
Here is some example that will replace all the '+' symbols in a UserName and NormalizedUserName JSON property of the Document table and replace it with an '@'. Simple example that I just composed because I ran my migrations on my prod website too quick and wanted to revert it.

``` SQL
-- Works with SQL Server

UPDATE Document
SET Content = JSON_MODIFY(Content, '$.UserName', REPLACE(JSON_VALUE(Content, '$.UserName'), '+', '@'))
WHERE Type = 'OrchardCore.Users.Models.User, OrchardCore.Users.Core'

UPDATE Document
SET Content = JSON_MODIFY(Content, '$.NormalizedUserName', REPLACE(JSON_VALUE(Content, '$.NormalizedUserName'), '+', '@'))
WHERE Type = 'OrchardCore.Users.Models.User, OrchardCore.Users.Core'

SELECT JSON_VALUE(Content, '$.UserName'), JSON_VALUE(Content, '$.NormalizedUserName')
FROM [dbo].[Document]
WHERE Type = 'OrchardCore.Users.Models.User, OrchardCore.Users.Core'
```

So here our eventual tool could search and replace instances of `JSON_MODIFY()` and `JSON_VALUE` functions to match it's proper related other RDB's functions ...

**Comment by jtkech** [comment]
@akim-green Just for info in the meantime

Does your `UpdateStoreInventory()` do a session commit?

If so, it may help to only do a session save and let the session be committed at the end of the current scope.

**Reply by akim-green** [reply]
I am not using any of the YesSQL functions except for Query<ContentItem,XPartIndex>() in UpdateStoreInventory(). There are several nested loops to check for the different components corresponding to a StoreInventoryItem. I make calls to update and publish content items within the loops however:

```
public async Task<string> UpdateStoreInfo()
        {
            var stores = (await _session.Query<ContentItem, StorePartIndex>().ListAsync()).ToList();

            foreach (ContentItem store in stores)
            {
                var storeInfo = GeoService.GetGeolocationData(store);
                await contentManager.LoadAsync(store);
                store.Alter<StorePart>(part => part.address = storeInfo.Address);
                await contentManager.UpdateAsync(store);
                await contentManager.PublishAsync(store);

            }

            return stores.Count + " stores were updated";
        }
```
would it be acceptable, or more performant if I looped through and publish separately. Are there any yessql functions or techniques that I could leverage? Thanks!

**Comment by jtkech** [comment, marked as answer]
About what i was talking that's okay, the above content manager methods don't call session commits, only session saves

But they call all related handlers, so if one handler does a query it will force a session flush which is a part of the commit work.

Hmm, normally you may need to also call `ValidateAsync()` but it may depend on your type / part

Only for testing try to just do a session save of each store in place of calling `UpdateAsync()` and `PublishAsync()`

**Reply by tropcicstefan** [reply]
@jtkech When you say query would 'force a session flush' does that mean that in those cases using dapper directly, without yessql's session, wouldn't have flushing as a drawback?

**Reply by jtkech** [reply]
@tropcicstefan @akim-green 

First, as @deanmarcussen said don't try to use the session in parallel e.g. with a task `WhenAll()`

Then i suggested to only use the ContentItem extensions to alter your already loaded stores content items, so not dapper as you already got your stores content items, then just do a session `Save()` yourself. This in place of calling `UpdateAsync()` and then `PublishAsync()`. This to prevent some session queries by some handlers.

But it was only for testing as anyway you need to publish your items, maybe the related handlers don't do any query, but at least `PublishAsync()` does a query to retrieve and unpublish the previous version.

Hmm, see in `DefaultContentManager.cs` what `PublishAsync()` does, maybe you could do a batch publish that would be called after the main loop. In this batch publish you could use a loop to call the handlers, update and save the previous version, update and save the new published version, this for each item, as done in the original `PublishAsync()`, but at the beginning you would have got all the previous versions in only one query.

Also, try it with the last dev branch (cloudsmith package feed), yesSql now supports batch updates.

**Reply by tropcicstefan** [reply]
@jtkech Thx for explanation, I was trying to understand YesSql more deeply.

My case is somewhat different. I need to update ContentItems, based on some other ContentItems i.e. querying db for current state and am starting to get problems with [deadlocks](https://github.com/OrchardCMS/OrchardCore/discussions/8445). So was thinking maybe avoiding .Query altogether is my way out.

**Reply by akim-green** [reply]
Thanks for the update @jtkech, I now have a clearer understanding of the steps you described.

However in my main loop I am also doing queries, so even when I remove the calls to Publish and replace them with session.save(), the session is still getting flushed I believe.

So I proceeded to remove all the queries from the main loop and instead do a query for all values that need to be searched. To be more specific, I have a list of UPCs I know are new to the store, but may not be new to the overall catalog. would like to do something like 
`                var CatalogProducts = (await _session.Query<ContentItem, ProductPartIndex>(index => productPartQueryString.Exists(item => item == index.FederalUpc)).ListAsync()).ToList();
`
but that does not work, I get a `variable 'index' of type 'App.Store.Indexes.ProductPartIndex' referenced from scope '', but it is not defined` error. Any tips on how I can accomplish such a query?

**Reply by jtkech** [reply]
There is an `IsIn()` yesSql extension `(index => index.SomePropertyOfTheIndex.IsIn(SomeEnumerableYouPopulated))`

**Reply by akim-green** [reply]
@jtkech Wonderful!! Thank you! That alone has greatly improved the performance of the function, now able to query and pull 300+ products very quickly. Will report once I finish refactoring.

**Reply by jtkech** [reply]
@akim-green Cool, happy to know it works well for your use case

**Reply by akim-green** [reply]
Thanks for all the suggestions. Just marking this as answer, as I was able to significantly increase performance by employing some of the techniques recommended by @jtkech. In my case all the items of interest were new Content Items, so in the main loop I saved them to a collection after initializing and altering, then passed the collection to function that created, updated, and published each item. This improved the update speed considerably, previously most updates would time out, now they mostly occur in less than 2 minutes for thousands of records, and outputs a report. Will update if/when I get around to trying the other recommended solutions.
